jax/core.py

```diff
@@ -210,11 +238,12 @@ def write(v, val):
   map(write, jaxpr.invars, args)
   for eqn in jaxpr.eqns:
     in_vals = map(read, eqn.invars)
-    if eqn.bound_subjaxpr:
-      subfuns = [lu.wrap_init(partial(eval_jaxpr, eqn.bound_subjaxpr, ()))]
+    call_jaxpr, params = extract_call_jaxpr(eqn.params)
+    if call_jaxpr:
```

Comment:
Would it be preferable to have a tag on the call primitives (e.g. by making them a subclass of Primitive, or else just having an explicit flag on the primitive itself), rather than handling them in a special way given a special parameter name? The latter feels a little fragile.

That is, how about:

```python
class CallPrimitive(Primitive): pass
```

and instead of checking `"call_jaxpr" in params` we just check `if type(eqn.primitive) is CallPrimitive`?

Reply:
You are right. I have chosen to add a .call_primitive attribute. I feel that adding another class is a more intrusive change, and unless we add specific methods to these classes it is not worth the cost.

Reply:
I think I’m with Matt here—we aren’t generally using classes as containers for methods as much as simply data types. CallPrimitive would be a fairly typical JAX use for a class.

Reply:
I don't think this point matters much, but purely out of curiosity, what cost are you referring to? I think writing `class CallPrimitive(Primitive): pass` is fewer lines of code (because we don't have to write separate lines like `remat_call_p.call_primitive = True` after `remat_call_p = core.Primitive('remat_call')`). A quick timing experiment suggests attribute lookup is faster than checking the type tag, though.

Reply:
I meant the "cognitive" cost, of having another class in the system. A boolean field seems to do the job.

Reply:
An alternative is to write core.Primitive(name, call_primitive, multiple_results)